### ISSUE TITLE
Support downloading of all data types

### ISSUE THREAD

**Issue by ffMathy** [ffMathy/Shapeshifter#275]
We should analyze @KoalaBear84's wonderful POC (thank you thank you THANK YOU!), and see what we can make of it.

Could help resolve #200 and #171.

**Comment by ffMathy** [comment]
@KoalaBear84's POC is in #200.

**Comment by ffMathy** [comment]
This has been fixed.

**Comment by KoalaBear84** [comment]
I still see too much incompatibility problems with the current approach. Shapeshifter should 'backup' the clipboard item completely as is, all available formats. Those formats are there exactly for compatibility with as much as possible other applications. It is also possible to create your own clipboard formats, as a software application.

Shapeshifter should be more like a clipboard viewer for the backed up items. Also, now Shapeshifter is altering the current clipboard item if you are holding CTRL + V, even the clipboard should not be altered because it is the current item. Therefore becoming a problem for unsupported formats. I know, that it could not restore it if it was the 'second' or more item, but it would at least not break the clipboard.

If Shapeshifter chooses to only pickup some of the formats available, as far as my current understanding, and treat the whole clipboard contents as such, it will never reach the 'native' compatibility.

I think my POC might work for all of this compatibility things, if I did know, at the time of writing, about the DIB and the specific Windows 'objects' for it, and restore the rest as is. The only thing unsupported would be the memory streams of file through RDP, which I might still see as a problem. BUT we could make a warning for the user that this item is unsupported and will be discarded if you copy another item.

TLDR; Serialize / backup ALL available formats, and make Shapeshifter more as a viewer / manager, not altering any of the formats, but restoring them as is

**Comment by KoalaBear84** [comment]
This would automatically fix issues like #290 or Excel things, well, everything :)

**Comment by ffMathy** [comment]
It already serializes everything and backs everything up. Which formats are you missing?

We can't operate on `DataObject` like in your POC due to bugs in the clipboard API. This was a lesson learned from Shapeshifter 5. It can produce some nasty issues. We must go the Win32 API way.

**Comment by ffMathy** [comment]
And #290 is not caused by this. See that issue for more information.

**Comment by KoalaBear84** [comment]
OK, that is something I didn't know. Do not have the knowledge of working with the clipboard like you have. Sorry, no hard feelings! :) Uhh, I don't know what to say anymore, then I would indeed need some more knowledge about Shapeshifter's internals, and the clipboard problems you experienced before.

I will check the problems and the solution for #290 in code to see if I understand. Thanks

**Comment by ffMathy** [comment]
Yeah likewise - no hard feelings ever, and sorry that Greenshot is still not working (although it will be working at the end of the day I think).

Here's a link for one of the most critical bugs with `DataObject` that Shapeshifter 5 had: https://social.msdn.microsoft.com/Forums/windows/en-US/ea562ff1-9558-425d-a30a-6a1809224edc/how-best-to-avoid-the-land-mines-planted-in-the-idataobject-returned-from-clipboardgetdataobject?forum=winforms

TL;DR: Sometimes (for instance when copying an image from Word), the `DataObject` will throw an `ExecutionEngineException` which is a special type of exception in .NET that can't be catched. **It's crazy.**

Shapeshifter 5 ended up primarily relying on `DataObject` and then for the formats it knew would crash, would rely on native methods. However, that was an ugly mix for clipboard contents where both types of format was included. Furthermore, it lead to its own set of issues with (for instance) streams, where because a `DataObject` had already been fetched, Shapeshifter couldn't fall back to native methods again for `Stream` types without corrupting the data.

If you think about it, `DataObject` is just a C# wrapper around native APIs. If they can do it, we can too. Heck, even their source codes are available! And we are almost there. Greenshot is almost ready, and most formats already convert one-to-one.

Knowing all of the above, do you understand my decision to make Shapeshifter 100% based on native "pure" methods?

**Comment by ffMathy** [comment]
Oh, and by the way, Excel copying should work also :+1:

**Comment by KoalaBear84** [comment]
Yes, I understand. So 100% skipping of `DataObject` and the C# `Clipboard` object, ok. Yes, offcourse, tested it myself, Excel was just an example 👍 :)

**Comment by ffMathy** [comment]
Perfect! :+1:

**Comment by ffMathy** [comment]
Given this new insight, I will close the issue again @KoalaBear84. Re-open if I misunderstood something, and in that case sorry for the inconvenience.
